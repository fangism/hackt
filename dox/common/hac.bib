% "hac.bib"
%	$Id: hac.bib,v 1.1 2007/01/24 04:14:47 fang Exp $

@STRING{ cacm = "Communications of the ACM" }

@ARTICLE{ ref:csp,
author = "C. Anthony R. Hoare",
title = "Communicating Sequential Processes",
journal = cacm,
volume = 21,
number = 8,
pages = "666--677",
year = 1978
}

