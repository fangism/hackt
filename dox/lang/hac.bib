% "hac.bib"
%	$Id: hac.bib,v 1.3 2006/10/06 03:30:09 fang Exp $

@STRING{ cacm = "Communications of the ACM" }

@ARTICLE{ ref:csp,
author = "C. Anthony R. Hoare",
title = "Communicating Sequential Processes",
journal = cacm,
volume = 21,
number = 8,
pages = "666--677",
year = 1978
}

