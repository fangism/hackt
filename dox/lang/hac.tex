% "hac.tex"

%	$Id: hac.tex,v 1.6 2006/06/03 05:52:31 fang Exp $

\documentclass[12pt]{book}

% for dependency-tracking macros
\input maketexdepend		% ../mk/maketexdepend
\maketexdependignorepattern{maketexdepend}

\usepackage{ifpdf}
\usepackage{verbatim}
\usepackage{makeidx}
\usepackage{amsthm}		% AMS-theorem

% COMPILE OPTIONS
\ifpdf
  \relax
\else
  \let\pdfoutput=\PDFoutput
\fi

\ifpdf
\usepackage[
	draft=false,
	pdftitle={The HAC Language}, 
	pdfauthor={David Fang}, 
	hyperfigures, 
	hyperindex, 
	bookmarks, 
	colorlinks
	]{hyperref}
\fi

\bibliographystyle{plain}

\makeindex

% PAGE FORMATTING
\setlength{\oddsidemargin}{.5in}
\setlength{\evensidemargin}{0in}
\setlength{\topmargin}{0in}
% \setlength{\topskip}{0in}
% \setlength{\footskip}{0in}
\setlength{\textwidth}{6in}
\setlength{\textheight}{8.5in}

% more sophisticated declaration of example environment
\newtheoremstyle{sectionexample}{\topsep}{\topsep}%
{\ttfamily}%	Body font
{}%	Indent amount (empty = no indent, \parindent = para indent)
{\bfseries}%	Thm head font
{}%	Punctuation after thm head
{\newline}%	Space after thm head (\newline = linebreak)
{\thmname{#1}\thmnumber{ #2}\thmnote{ #3}}%	Thm head spec
\theoremstyle{sectionexample}
\newtheorem{sectionexample}{Example}[section]

% kind of dirty space-hacking, but modular
\newenvironment{example}[1]
  {\begin{sectionexample}[#1]\ \\ \vspace{-2.5em}\begin{quote}}
  {\end{quote}\end{sectionexample}}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% \maketexdependignore{hacdef.tex}
\maketexdependignorepattern{hacdef}
\input hacdef		% from $(srcdir)/../common/hacdef.tex

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% end of preamble
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{document}

\title{The HAC Language}
\author{David Fang}

\ifpdf\pdfbookmark[0]{Title Page}{title}\fi
\maketitle

% trying to correct odd even margins...
\begin{comment}
\clearpage
\centerline{This page is intentionally left blank.}
\clearpage
\centerline{So is this page}
\clearpage
\end{comment}

\ifpdf\pdfbookmark[0]{Table of Contents}{toc}\fi
\tableofcontents

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\input chapters/preface
\input chapters/goals
\input chapters/intro
\input chapters/types
\input chapters/expressions
\input chapters/arrays
\input chapters/processes
\input chapters/channels
\input chapters/datatypes
\input chapters/namespaces
\input chapters/templates
% \input chapters/instances
\input chapters/connections
\input chapters/typedefs
% \input chapters/control	% loops and conditionals
\input chapters/linkage
\input chapters/CHP
\input chapters/PRS
\input chapters/SPEC

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\ifpdf\pdfbookmark[0]{Bibliography}{bibliography}\fi
\maketexdependadd{hac.bib}
\bibliography{hac}

\clearpage
\ifpdf\pdfbookmark[0]{Index}{index}\fi
\printindex

\end{document}


