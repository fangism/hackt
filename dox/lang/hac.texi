\input texinfo	@c -*-texinfo-*-
@c %**start of header
@setfilename hac.info
@settitle The HAC Language
@c `on' is the default
@setchapternewpage odd
@c for two-sided book-style printing
@c %**end of header

@c "hac.texi"
@c $Id: hac.texi,v 1.6 2008/11/28 22:07:05 fang Exp $

@dircategory Hierarchical Asynchronous Circuit Kompiler Toolkit (HACKT)
@direntry
* Hierarchical Asynchronous Circuits: (hac).  The HAC Language Reference.
@end direntry

@include hacmacros.texi

@copying
This manual describes the HAC language specification.

This document can also be found online at @homepagesubref{hac}.

The main project home page is @homepageref{}.

Source code for this document can be found at @sourcerepo{dox/lang}.

@copyrightnotice{2007}

@quotation
Published by ...
 
Permission is hereby granted to distribute this document.
@end quotation
@end copying

@c Title page and Copyright
@titlepage
@title The HAC Languge
@subtitle A boring technical document
@author David Fang
@page
@vskip 0pt plus 1filll
@insertcopying
@end titlepage

@c ****************************** contents ************************************
@shortcontents
@contents

@ignore
@iftex
@node List of Figures
@unnumbered List of Figures
@listoffloats Figure
@end iftex
@end ignore

@c ****************************** Top *****************************************
@ifnottex
@node Top
@top The HAC Language
@insertcopying
@end ifnottex

@c @copying is recommended here as well
@menu
* Preface::
* Goals::
* Introduction:: Introduction to HAC.
* Types::
* Expressions::
* Arrays::
* Processes::
* Channels::
* Datatypes::
* Namespaces::
* Templates::
* Connections::
* Attributes::
* Typedefs::
* Linkage::
* CHP::
* PRS::
* SPEC Directives::
* Run-time atomic expressions::
* Keywords::
* Grammar::
* Function Index::
* Concept Index::
@end menu

@c ****************************** chapter *************************************
@include chapters/preface.texi
@include chapters/goals.texi
@include chapters/intro.texi
@include chapters/types.texi
@include chapters/expressions.texi
@include chapters/arrays.texi
@include chapters/processes.texi
@include chapters/channels.texi
@include chapters/datatypes.texi
@include chapters/namespaces.texi
@include chapters/templates.texi
@include chapters/connections.texi
@include chapters/attributes.texi
@include chapters/typedefs.texi
@include chapters/linkage.texi
@include chapters/CHP.texi
@include chapters/PRS.texi
@include chapters/SPEC.texi
@include chapters/RTE.texi
@include grammar/keywords.texi
@include grammar/grammar.texi

@c ****************************** index ***************************************
@node Function Index
@unnumbered Function Index
@printindex fn

@node Concept Index
@unnumbered Concept Index
@printindex cp

@c ****************************** finis ***************************************
@bye

