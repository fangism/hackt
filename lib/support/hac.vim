" Vim syntax file
" Language:	HAC
" Maintainer:	CARLOS TADEO ORTEGA OTERO, DAVID FANG
" Last Change:	2010-01-13


" A bunch of useful C keywords
syn keyword	hacImport  	import
syn match	hacImport  	"#FILE"
syn match	hacImport  	"%{"
syn match	hacImport  	"%}"
syn keyword	cStatement  	template export open 
" syn keyword	cLabel		case default
syn keyword	cLabel		defproc deftype defchan namespace
syn keyword	cLabel		spec chp prs hse rte
syn keyword	cConditional	else skip
"syn keyword	cConditional	if else switch
"syn keyword	cRepeat	  	for while	

syn keyword	cTodo		contained TODO FIXME XXX BUG

" cCommentGroup allows adding matches for special things in comments
syn cluster	cCommentGroup	contains=cTodo

" String and Character constants
" Highlight special characters (those which have a backslash) differently
syn match	cSpecial	display contained "\\\(x\x\+\|\o\{1,3}\|.\|$\)"
if !exists("c_no_utf")
  syn match	cSpecial	display contained "\\\(u\x\{4}\|U\x\{8}\)"
endif
if exists("c_no_cformat")
  syn region	cString		start=+L\="+ skip=+\\\\\|\\"+ end=+"+ contains=cSpecial
  " cCppString: same as cString, but ends at end of line
  syn region	cCppString	start=+L\="+ skip=+\\\\\|\\"\|\\$+ excludenl end=+"+ end='$' contains=cSpecial
else
  syn match	cFormat		display "%\(\d\+\$\)\=[-+' #0*]*\(\d*\|\*\|\*\d\+\$\)\(\.\(\d*\|\*\|\*\d\+\$\)\)\=\([hlL]\|ll\)\=\([diuoxXDOUfeEgGcCsSpn]\|\[\^\=.[^]]*\]\)" contained
  syn match	cFormat		display "%%" contained
  syn region	cString		start=+L\="+ skip=+\\\\\|\\"+ end=+"+ contains=cSpecial,cFormat
  " cCppString: same as cString, but ends at end of line
  syn region	cCppString	start=+L\="+ skip=+\\\\\|\\"\|\\$+ excludenl end=+"+ end='$' contains=cSpecial,cFormat
endif

syn match	cCharacter	"L\='[^\\]'"
syn match	cCharacter	"L'[^']*'" contains=cSpecial
if exists("c_gnu")
  syn match	cSpecialError	"L\='\\[^'\"?\\abefnrtv]'"
  syn match	cSpecialCharacter "L\='\\['\"?\\abefnrtv]'"
else
  syn match	cSpecialError	"L\='\\[^'\"?\\abfnrtv]'"
  syn match	cSpecialCharacter "L\='\\['\"?\\abfnrtv]'"
endif
syn match	cSpecialCharacter display "L\='\\\o\{1,3}'"
syn match	cSpecialCharacter display "'\\x\x\{1,2}'"
syn match	cSpecialCharacter display "L'\\x\x\+'"

"when wanted, highlight trailing white space
if exists("c_space_errors")
  if !exists("c_no_trail_space_error")
    syn match	cSpaceError	display excludenl "\s\+$"
  endif
  if !exists("c_no_tab_space_error")
    syn match	cSpaceError	display " \+\t"me=e-1
  endif
endif

"catch errors caused by wrong parenthesis and brackets
" also accept <% for {, %> for }, <: for [ and :> for ] (C99)
syn cluster	cParenGroup	contains=cParenError,cIncluded,cSpecial,cCommentSkip,cCommentString,cComment2String,@cCommentGroup,cCommentStartError,cUserCont,cUserLabel,cBitField,cCommentSkip,cOctalZero,cCppOut,cCppOut2,cCppSkip,cFormat,cNumber,cFloat,cOctal,cOctalError,cNumbersCom
if exists("c_no_bracket_error")
  syn region	cParen		transparent start='(' end=')' contains=ALLBUT,@cParenGroup,cCppParen,cCppString

  " cCppParen: same as cParen but ends at end-of-line; used in cDefine
  syn region	cCppParen	transparent start='(' skip='\\$' excludenl end=')' end='$' contained contains=ALLBUT,@cParenGroup,cParen,cString
  syn match	cParenError	display ")"
  syn match	cErrInParen	display contained "[{}]\|<%\|%>"
else
  syn region	cParen		transparent start='(' end=')' contains=ALLBUT,@cParenGroup,cCppParen,cErrInBracket,cCppBracket,cCppString
  " cCppParen: same as cParen but ends at end-of-line; used in cDefine
  syn region	cCppParen	transparent start='(' skip='\\$' excludenl end=')' end='$' contained contains=ALLBUT,@cParenGroup,cErrInBracket,cParen,cBracket,cString
  syn match	cParenError	display "[\])]"
  syn match	cErrInParen	display contained "[\]]\|<%\|%>"
  syn region	cBracket	transparent start='\[\|<:' end=']\|:>' contains=ALLBUT,@cParenGroup,cErrInParen,cCppParen,cCppBracket,cCppString
  " cCppBracket: same as cParen but ends at end-of-line; used in cDefine
  syn region	cCppBracket	transparent start='\[\|<:' skip='\\$' excludenl end=']\|:>' end='$' contained contains=ALLBUT,@cParenGroup,cErrInParen,cParen,cBracket,cString
  syn match	cErrInBracket	display contained "[)]\|<%\|%>"
endif

"integer number, or floating point number without a dot and with "f".
syn case ignore
syn match	cNumbers	display transparent "\<\d\|\.\d" contains=cNumber,cFloat,cOctalError,cOctal
" Same, but without octal error (for comments)
syn match	cNumbersCom	display contained transparent "\<\d\|\.\d" contains=cNumber,cFloat,cOctal
syn match	cNumber		display contained "\d\+\(u\=l\{0,2}\|ll\=u\)\>"
"hex number
syn match	cNumber		display contained "0x\x\+\(u\=l\{0,2}\|ll\=u\)\>"
" Flag the first zero of an octal number as something special
syn match	cOctal		display contained "0\o\+\(u\=l\{0,2}\|ll\=u\)\>" contains=cOctalZero
syn match	cOctalZero	display contained "\<0"
syn match	cFloat		display contained "\d\+f"
"floating point number, with dot, optional exponent
syn match	cFloat		display contained "\d\+\.\d*\(e[-+]\=\d\+\)\=[fl]\="
"floating point number, starting with a dot, optional exponent
syn match	cFloat		display contained "\.\d\+\(e[-+]\=\d\+\)\=[fl]\=\>"
"floating point number, without dot, with exponent
syn match	cFloat		display contained "\d\+e[-+]\=\d\+[fl]\=\>"
if !exists("c_no_c99")
  "hexadecimal floating point number, optional leading digits, with dot, with exponent
  syn match	cFloat		display contained "0x\x*\.\x\+p[-+]\=\d\+[fl]\=\>"
  "hexadecimal floating point number, with leading digits, optional dot, with exponent
  syn match	cFloat		display contained "0x\x\+\.\=p[-+]\=\d\+[fl]\=\>"
endif

" flag an octal number with wrong digits
syn match	cOctalError	display contained "0\o*[89]\d*"
syn case match

if exists("c_comment_strings")
  " A comment can contain cString, cCharacter and cNumber.
  " But a "*/" inside a cString in a cComment DOES end the comment!  So we
  " need to use a special type of cString: cCommentString, which also ends on
  " "*/", and sees a "*" at the start of the line as comment again.
  " Unfortunately this doesn't very well work for // type of comments :-(
  syntax match	cCommentSkip	contained "^\s*\*\($\|\s\+\)"
  syntax region cCommentString	contained start=+L\=\\\@<!"+ skip=+\\\\\|\\"+ end=+"+ end=+\*/+me=s-1 contains=cSpecial,cCommentSkip
  syntax region cComment2String	contained start=+L\=\\\@<!"+ skip=+\\\\\|\\"+ end=+"+ end="$" contains=cSpecial
  syntax region  cCommentL	start="//" skip="\\$" end="$" keepend contains=@cCommentGroup,cComment2String,cCharacter,cNumbersCom,cSpaceError
  syntax region cComment	matchgroup=cCommentStart start="/\*" end="\*/" contains=@cCommentGroup,cCommentStartError,cCommentString,cCharacter,cNumbersCom,cSpaceError
else
  syn region	cCommentL	start="//" skip="\\$" end="$" keepend contains=@cCommentGroup,cSpaceError
  syn region	cComment	matchgroup=cCommentStart start="/\*" end="\*/" contains=@cCommentGroup,cCommentStartError,cSpaceError
endif
" keep a // comment separately, it terminates a preproc. conditional
syntax match	cCommentError	display "\*/"
syntax match	cCommentStartError display "/\*"me=e-1 contained

syn match	cOperator	display "->\|=>\|#>"
syn match	cOperator	display "&\||\|\~"
syn match	cOperator	display "\.\."
syn match	cOperator	display "\\$"

" syn keyword	cOperator	sizeof
syn keyword	cType  		chan pint pbool preal pstring
syn keyword	cType  		int bool ebool eint real string
" syn keyword	cOperator	typeof
" syn keyword	cType		char void
" syn keyword	cType		signed unsigned float double

syn keyword	cStructure	enum typedef
" syn keyword	cStructure	struct union
" syn keyword	cStorageClass	static register auto volatile extern const
" syn keyword	cStorageClass	inline
" syn keyword cConstant __LINE__ __FILE__ __DATE__ __TIME__
syn keyword	cConstant	true false

" Accept %: for # (C99)
syn region	cPreCondit	start="^\s*\(%:\|#\)\s*\(if\|ifdef\|ifndef\|elif\)\>" skip="\\$" end="$" end="//"me=s-1 contains=cComment,cCppString,cCharacter,cCppParen,cParenError,cNumbers,cCommentError,cSpaceError
syn match	cPreCondit	display "^\s*\(%:\|#\)\s*\(else\|endif\)\>"
syn match	hacRuleAttrib	display "^\s*\[.\{-}\]"
if !exists("c_no_if0")
  syn region	cCppOut		start="^\s*\(%:\|#\)\s*if\s\+0\+\>" end=".\@=\|$" contains=cCppOut2
  syn region	cCppOut2	contained start="0" end="^\s*\(%:\|#\)\s*\(endif\>\|else\>\|elif\>\)" contains=cSpaceError,cCppSkip
  syn region	cCppSkip	contained start="^\s*\(%:\|#\)\s*\(if\>\|ifdef\>\|ifndef\>\)" skip="\\$" end="^\s*\(%:\|#\)\s*endif\>" contains=cSpaceError,cCppSkip
endif
syn region	cIncluded	display contained start=+"+ skip=+\\\\\|\\"+ end=+"+
syn match	cIncluded	display contained "<[^>]*>"
syn match	cInclude	display "^\s*\(%:\|#\)\s*import\>\s*["<]" contains=cIncluded
"syn match cLineSkip	"\\$"
syn cluster	cPreProcGroup	contains=cPreCondit,cIncluded,cInclude,cDefine,cErrInParen,cErrInBracket,cUserLabel,cSpecial,cOctalZero,cCppOut,cCppOut2,cCppSkip,cFormat,cNumber,cFloat,cOctal,cOctalError,cNumbersCom,cString,cCommentSkip,cCommentString,cComment2String,@cCommentGroup,cCommentStartError,cParen,cBracket
syn region	cDefine		start="^\s*\(%:\|#\)\s*\(include\|undef\)\>" skip="\\$" end="$" end="//"me=s-1 contains=ALLBUT,@cPreProcGroup
syn region	cPreProc	start="^\s*\(%:\|#\)\s*\(pragma\>\|line\>\|warning\>\|warn\>\|error\>\)" skip="\\$" end="$" keepend contains=ALLBUT,@cPreProcGroup

" Highlight User Labels
syn cluster	cMultiGroup	contains=cIncluded,cSpecial,cCommentSkip,cCommentString,cComment2String,@cCommentGroup,cCommentStartError,cUserCont,cUserLabel,cBitField,cOctalZero,cCppOut,cCppOut2,cCppSkip,cFormat,cNumber,cFloat,cOctal,cOctalError,cNumbersCom,cCppParen,cCppBracket,cCppString
"syn region	cMulti		transparent start='?' skip='::' end=':' contains=ALLBUT,@cMultiGroup
" Avoid matching foo::bar() in C++ by requiring that the next char is not ':'
syn cluster	cLabelGroup	contains=cUserLabel
syn match	cUserCont	display "^\s*\I\i*\s*:$" contains=@cLabelGroup
syn match	cUserCont	display ";\s*\I\i*\s*:$" contains=@cLabelGroup
syn match	cUserCont	display "^\s*\I\i*\s*:[^:]"me=e-1 contains=@cLabelGroup
syn match	cUserCont	display ";\s*\I\i*\s*:[^:]"me=e-1 contains=@cLabelGroup

syn match	cUserLabel	display "\I\i*" contained

" Avoid recognizing most bitfields as labels
syn match	cBitField	display "^\s*\I\i*\s*:\s*[1-9]"me=e-1
syn match	cBitField	display ";\s*\I\i*\s*:\s*[1-9]"me=e-1

if exists("c_minlines")
  let b:c_minlines = c_minlines
else
  if !exists("c_no_if0")
    let b:c_minlines = 50	" #if 0 constructs can be long
  else
    let b:c_minlines = 15	" mostly for () constructs
  endif
endif
exec "syn sync ccomment cComment minlines=" . b:c_minlines

" Define the default highlighting.
" For version 5.7 and earlier: only when not done already
" For version 5.8 and later: only when an item doesn't have highlighting yet
if version >= 508 || !exists("did_c_syn_inits")
  if version < 508
    let did_c_syn_inits = 1
    command -nargs=+ HiLink hi link <args>
  else
    command -nargs=+ HiLink hi def link <args>
  endif

  HiLink cFormat		cSpecial
  HiLink cCppString		cString
  HiLink cCommentL		cComment
  HiLink cCommentStart		cComment
  HiLink cLabel			Label
  HiLink cUserLabel		Label
  HiLink cConditional		Conditional
  HiLink cRepeat		Repeat
  HiLink cCharacter		Character
  HiLink cSpecialCharacter	cSpecial
  HiLink cNumber		Number
  HiLink cOctal			Number
  HiLink cOctalZero		PreProc	 " link this to Error if you want
  HiLink cFloat			Float
  HiLink cOctalError		cError
  HiLink cParenError		cError
  HiLink cErrInParen		cError
  HiLink cErrInBracket		cError
  HiLink cCommentError		cError
  HiLink cCommentStartError	cError
  HiLink cSpaceError		cError
  HiLink cSpecialError		cError
  HiLink cOperator		Operator
  HiLink cStructure		Structure
  HiLink cStorageClass		StorageClass
  HiLink cInclude		Include
  HiLink cPreProc		PreProc
  HiLink hacImport		PreProc
  HiLink cDefine		Macro
  HiLink cIncluded		cString
  HiLink cError			Error
  HiLink cStatement		Statement
  HiLink cPreCondit		PreCondit
  HiLink hacRuleAttrib		cLabel
  HiLink cType			Type
  HiLink cConstant		Constant
  HiLink cCommentString		cString
  HiLink cComment2String	cString
  HiLink cCommentSkip		cComment
  HiLink cString		String
  HiLink cComment		Comment
  HiLink cSpecial		SpecialChar
  HiLink cTodo			Todo
  HiLink cCppSkip		cCppOut
  HiLink cCppOut2		cCppOut
  HiLink cCppOut		Comment

  delcommand HiLink
endif

let b:current_syntax = "c"

" vim: ts=8
