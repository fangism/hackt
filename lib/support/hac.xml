<!-- "hac.xml" -->
<!-- $Id: hac.xml,v 1.1 2008/03/25 20:00:55 fang Exp $ -->
<!-- contributed/maintained by Carlos Tadeo Ortega Otero <cto3@csl.cornell.edu> -->
<!-- also usable for the ACT dialect -->

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE language SYSTEM "language.dtd">
<language name="Act" version="1.0" kateversion="2.4" section="Sources" extensions="*.act" mimetype="text" license="LGPL" author="Carlos Tadeo Ortega Otero (cto3@cornell.edu)">
	<highlighting>
		<list name="keywords">
			<item> import </item>
			<item> namespace </item>
			<item> open </item>
			<item> as </item>
			<item> template </item>
			<item> define </item>
			<item> defchan </item>
			<item> deftype </item>
			<item> defproc </item>
			<item> typedef </item>
			<item> chp </item>
			<item> hse </item>
			<item> prs </item>
			<item> rte </item>
			<item> spec </item>
			<item> tree </item>
			<item> subckt </item>
			<item> skip </item>
			<item> else </item>
			<item> log </item>
			<item> send </item>
			<item> recv </item>
			<item> set </item>
			<item> get </item>
			<item> enum </item>
			<item> chan </item>
			<item> true </item>
			<item> false </item>
			<item> extern </item>
			<item> static </item>
			<item> export </item>
		</list>
		<list name="types">
			<item> export </item>
			<item> e1of </item>
			<item> bool </item>
			<item> int </item>
			<item> ebool </item>
			<item> eint </item>
			<item> pbool </item>
			<item> pint </item>
			<item> real </item>
			<item> preal </item>
		</list>
		<contexts>
			<context attribute="Normal Text" lineEndContext="#stay" name="Normal">
				<keyword attribute="Keyword" context="#stay" String="keywords"/>
				<keyword attribute="Data Type" context="#stay" String="types"/>

				<HlCOct attribute="Octal" context="#stay"/>
				<HlCHex attribute="Hex" context="#stay"/>
				<Int attribute="Decimal" context="#stay">
					<StringDetect attribute="Decimal" context="#stay" String="ULL" insensitive="TRUE"/>
					<StringDetect attribute="Decimal" context="#stay" String="LUL" insensitive="TRUE"/>
					<StringDetect attribute="Decimal" context="#stay" String="LLU" insensitive="TRUE"/>
					<StringDetect attribute="Decimal" context="#stay" String="UL" insensitive="TRUE"/>
					<StringDetect attribute="Decimal" context="#stay" String="LU" insensitive="TRUE"/>
					<StringDetect attribute="Decimal" context="#stay" String="LL" insensitive="TRUE"/>
					<StringDetect attribute="Decimal" context="#stay" String="U" insensitive="TRUE"/>
					<StringDetect attribute="Decimal" context="#stay" String="L" insensitive="TRUE"/>
				</Int>
				<HlCChar attribute="Char" context="#stay"/>
				<RegExpr attribute="Decimal" context="#stay" String="//\s*BEGIN.*$" beginRegion="Region1"/>
				<RegExpr attribute="Decimal" context="#stay" String="//\s*END.*$" endRegion="Region1"/>
				<DetectChar attribute="String" context="String" char="&quot;"/>
				<Detect2Chars attribute="Comment" context="Commentar 1" char="/" char1="/"/>
				<Detect2Chars attribute="Comment" context="Commentar 2" char="/" char1="*" beginRegion="Comment"/>
				<DetectChar attribute="Symbol" context="#stay" char="{" beginRegion="Brace1"/>
				<DetectChar attribute="Symbol" context="#stay" char="}" endRegion="Brace1"/>
				<RegExpr attribute="Keyword" context="#stay" String="\.{3,3}\s+" />
				<RegExpr attribute="Keyword" context="StaticImports" String="\b(import\s+static)\b" />
				<RegExpr attribute="Keyword" context="Imports" String="\b(package|import)\b" />
				<RegExpr attribute="Function" context="#stay" String="\b[_\w][_\w\d]*(?=[\s]*(/\*\s*\d+\s*\*/\s*)?[(])" />
				<RegExpr attribute="Symbol" context="Member" String="[.]{1,1}" />
				<AnyChar attribute="Symbol" context="#stay" String=":!%&amp;()+,-/.*&lt;=&gt;?[]|~^&#59;"/>
			</context>
			<context attribute="String" lineEndContext="#pop" name="String">
				<LineContinue attribute="String" context="#stay"/>
				<HlCStringChar attribute="String Char" context="#stay"/>
				<DetectChar attribute="String" context="#pop" char="&quot;"/>
			</context>
			<context attribute="Normal Text" lineEndContext="#pop" name="Member" fallthrough="true" fallthroughContext="#pop">
				<RegExpr attribute="Function" context="#pop" String="\b[_a-zA-Z]\w*(?=[\s]*)" />
			</context>
			<context attribute="Normal Text" lineEndContext="#pop" name="StaticImports">
				<RegExpr attribute="StaticImports" context="#pop" String="\s*.*;" />
			</context>
			<context attribute="Normal Text" lineEndContext="#pop" name="Imports">
				<RegExpr attribute="Imports" context="#pop" String="\s*.*;" />
			</context>
			<context attribute="Comment" lineEndContext="#pop" name="Commentar 1"/>
			<context attribute="Comment" lineEndContext="#stay" name="Commentar 2">
				<Detect2Chars attribute="Comment" context="#pop" char="*" char1="/" endRegion="Comment"/>
			</context>
		</contexts>
		<itemDatas>
			<itemData name="Normal Text"  defStyleNum="dsNormal"/>
			<itemData name="Keyword"      defStyleNum="dsNormal" bold="1" italic="1" color="#800080"/>
			<itemData name="Function"     defStyleNum="dsFunction"/>
			<itemData name="StaticImports"      defStyleNum="dsKeyword" color="#800080" selColor="#FFFFFF" bold="0" italic="0"/>
			<itemData name="Imports"      defStyleNum="dsKeyword" color="#808000" selColor="#FFFFFF" bold="0" italic="0"/>
			<itemData name="Data Type"    defStyleNum="dsDataType"/>
			<itemData name="Decimal"      defStyleNum="dsDecVal"/>
			<itemData name="Octal"        defStyleNum="dsBaseN"/>
			<itemData name="Hex"          defStyleNum="dsBaseN"/>
			<itemData name="Float"        defStyleNum="dsFloat"/>
			<itemData name="Char"         defStyleNum="dsChar"/>
			<itemData name="String"       defStyleNum="dsString"/>
			<itemData name="String Char"  defStyleNum="dsChar"/>
			<itemData name="Comment"      defStyleNum="dsComment"/>
			<itemData name="Symbol"       defStyleNum="dsNormal"/>
		</itemDatas>
	</highlighting>
	<general>
		<comments>
			<comment name="singleLine" start="//"/>
			<comment name="multiLine" start="/*" end="*/"/>
		</comments>
		<keywords casesensitive="1"/>
	</general>
</language>

