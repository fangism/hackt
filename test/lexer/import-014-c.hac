// "import-014-c.hac"
// contributed by: Clint Kelly
// reduced by: David Fang
// purpose: testing bug fix in lexer's import mechanism
//	used to assert-fail during file-stack dump
// P.S: point of failure had to be changed because we added macro support
import "import-014-a.hac";
import "import-014-b.hac";

defproc problem (e1of2 Din; e1of2 Dout; pbit pb)
{
	bool p = pb.t;
	bool _p = pb.f;

	// DEATH HERE on macro-like construct in prs
	prs {
		foo+
		// foo(_p, Din.e, Dout.e)
	}
}

