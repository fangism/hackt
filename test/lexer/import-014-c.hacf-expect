// "import-014-c.hac"
// contributed by: Clint Kelly
// reduced by: David Fang
// purpose: testing bug fix in lexer's import mechanism
//	used to assert-fail during file-stack dump
// P.S: point of failure had to be changed because we added macro support
#FILE "import-014-a.hac" %{
// "import-014-a.hac"
// contributed by: Clint Kelly
// reduced by: David Fang

defproc ch1of2 (bool d[2]; bool t,f)
{
	d[0]=f;
	d[1]=t;
}

defproc pbit (bool d[2]; bool t,f)
{
	d[0]=f;
	d[1]=t;
}

defproc e1of2 (bool d[2]; bool t,f,e)
{
	d[0]=f;
	d[1]=t;
}


%}	// #FILE "import-014-a.hac"
#FILE "import-014-b.hac" %{
/**
	"import-014-b.hac"
	contributed by: Clint Kelly
	reduced by: David Fang
*/



defproc srambit (pbit pb; ch1of2 b, w)
{
	prs {
		w & b.t -> pb.t+				
		w & b.t -> pb.f-				
		w & b.f -> pb.f+				
		w & b.f -> pb.t-				
	}
}

%}	// #FILE "import-014-b.hac"

defproc problem (e1of2 Din; e1of2 Dout; pbit pb)
{
	bool p = pb.t;
	bool _p = pb.f;

	// DEATH HERE on macro-like construct in prs
	prs {
		foo+
		// foo(_p, Din.e, Dout.e)
	}
}

