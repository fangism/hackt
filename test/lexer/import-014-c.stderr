At: "import-014-c.hac":17:
parse error: syntax error
parser stacks:
state	value
#STATE#	(imported-root-list) [7:1--8:27]
#STATE#	(process-prototype) [10:1..48]
#STATE#	{ [11:1]
#STATE#	list<(def-body-item)>: (instance-decl) ... [12:2--13:15]
#STATE#	keyword: prs [16:2..4]
#STATE#	(null) 
#STATE#	{ [16:6]
#STATE#	(prs-literal) [17:3..5]
#STATE#	+ [17:6]
in state #STATE#, possible rules are:
	prs_literal : prs_literal_base . prs_literal_params_in_angles_optional  (#RULE#)
	prs_literal_params_in_angles_optional : .  (#RULE#)
acceptable tokens are: 
	'&' (reduce)
	'(' (reduce)
	')' (reduce)
	',' (reduce)
	'<' (shift)
	'|' (reduce)
	'}' (reduce)
	IMPLIES (reduce)
	RARROW (reduce)
	HASH_ARROW (reduce)
